feat(secureapplication):initial implementation of Secure Application

sgulraja opened the pull request:
feat(secureapplication):initial implementation of Secure Application SOAR connector
    
    Add first version of Secure Application SOAR app including:
    - App JSON with configuration and actions
    - Action handlers for creating, updating, deleting, and listing policies
    - Connectivity check
    
    This connector allows Phantom to interface with Cisco Secure Application's API to manage policies.

sgulraja committed:
feat(secureapplication):initial implementation of Secure Application SOAR connector

    Add first version of Secure Application SOAR app including:
    - App JSON with configuration and actions
    - Action handlers for creating, updating, deleting, and listing policies
    - Connectivity check

    This connector allows Phantom to interface with Cisco Secure Application's API to manage policies.

sgulraja committed:
fix: Use "Cisco" instead of "Cisco Systems" in Vendor

sgulraja committed:
fix: Changed the output of most actions to display more info

sgulraja committed:
fix:Remove some columns from output

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+# limitations under the License.
+#!/usr/bin/python
+# -----------------------------------------
+# Phantom sample App Connector python file
```
Should this python shebang and sample comment be here? Probably remove

sgulraja replied:
fixed

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
 import phantom.app as phantom
+
+# Usage of the consts file is recommended
+# from secureapplication_consts import *
```
I noticed there is no consts file present did you want to make that and move any constants over?

sgulraja replied:
added a couple constants

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+
+        password = getpass.getpass("Password: ")
+
+    if username and password:
```
What is happening here with this session stuff? Do you need this or was this just for debugging? I feel you could clean this up and remove the username, password, and session stuff here

sgulraja replied:
done

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+        except Exception as e:
+            return RetVal(action_result.set_status(phantom.APP_ERROR, f"Unable to parse JSON response. Error: {e!s}"), None)
+
+        # Please specify the status codes here
```
Old comment? Or are there still supposed to be changes here?

sgulraja replied:
old comment. removed

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+        self.save_progress(f"In action handler for: {self.get_action_identifier()}")
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        endpoint = "/controller/argento/public-api/v1/libraries"
```
Endpoints for example are great to keep in a consts file, easy to change if the API changes etc...

sgulraja replied:
Moved a couple constant paths to consts.py

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+
+        if missing:
+            missing_str = ", ".join(missing)
+            return action_result.set_status(phantom.APP_ERROR, "Missing required parameter(s): " + missing_str)
```
Use format string when you can (bit cleaner)
```suggestion
return action_result.set_status(phantom.APP_ERROR, f"Missing required parameter(s): {missing_str}")
```

grokas-splunk replied:
Try to apply this everywhere 👍

sgulraja replied:
done. thanks for the suggestion. Had missed doing this at a couple places

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+
+        self.debug_print("action_id", self.get_action_identifier())
+
+        if action_id == "create_new_policy":
```
An action map can be a cleaner routing solution when dealing with many actions (like here):

[Example action map for reference](https://github.com/splunk-soar-connectors/ciscosma/blob/main/ciscosma_connector.py#L1159-L1177)

sgulraja replied:
done

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+        action_result.update_summary({"total_rules": len(rules)})
+        return action_result.set_status(status, message)
+
+    def handle_action(self, param):
```
This function should be below all your actions (should be just above `main`)

sgulraja replied:
done

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+        elif rule_type == "process":
+            match_field = "targetMatch"
+        else:
+            return action_result.set_status(phantom.APP_ERROR, f"Unsupported rule type: {rule_type}")
```
Cleaner

```suggestion
if rule_type == "stack trace":
    match_field = "stackMatch"
elif rule_type in ("filename", "hostname", "process"):
    match_field = "targetMatch"
else:
    return action_result.set_status(phantom.APP_ERROR, f"Unsupported rule type: {rule_type}")
```

sgulraja replied:
done

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,923 @@
+        endpoint = f"/controller/argento/public-api/v1/policyConfigs/{policy_id}"
+        headers = self._get_rest_api_headers(token=self._token)
+
+        self.debug_print(f"Sending payload: {json.dumps(updated_policy, indent=2)}")
```
Think we’ve got more debug_prints than we need in the file, but might be worth trimming the ones that don’t add much to keep the code cleaner

sgulraja replied:
Removed some but they have been very helpful when debugging internally

grokas-splunk reviewed (changes requested):
Overall looking good with just some needed changes here and there. Remember to add your `manual_readme_content.md` too 👍

sgulraja committed:
fix/style: Make changes as per review comments

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,825 @@
-    SECUREAPPLICATION_ERR_CONNECTIVITY_TEST,
-    SECUREAPPLICATION_SUCC_CONNECTIVITY_TEST,
-)
+from secureapplication_consts import *
```
Think it's better not to use * and actually specifiy the imports from the const file (better for tracking)

sgulraja replied:
done

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,825 @@
+        self.save_progress(f"In action handler for: {self.get_action_identifier()}")
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        required_params = ["type", "application_id", "tier_id", "default_action", "enable_policy"]
```
Could something like this go in the consts file?

sgulraja replied:
These are params specified in the json file for this action. This is more readable

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,825 @@
+        action_result.update_summary({"total_rules": len(rules)})
+        return action_result.set_status(status, message)
+
     def initialize(self):
```
Probably move this function elsewhere (to the bottom perhaps?) seems strange to just be between a bunch of actions

sgulraja replied:
done

grokas-splunk commented on secureapplication_connector.py:
```diff
@@ -8,138 +6,825 @@
+
+        return phantom.APP_SUCCESS
+
+    def finalize(self):
```
Same with this function

grokas-splunk reviewed (changes requested):
Few more comments

sgulraja committed:
fix: Add the manual readme

sgulraja committed:
style/fix:
Changes as per review comments
Fixed  a param description
Changed a return status on failure
Changed the manual doc as per the review from Ada Ndili

sgulraja committed:
fix: Add support for creating headers for http transaction policy, which was missed earlier.

grokas-splunk committed:
chore: release notes added
